src/java_bytecode/java_object_factory.cpp

```diff
@@ -753,12 +753,21 @@ void java_object_factoryt::gen_nondet_pointer_init(
   const update_in_placet &update_in_place)
 {
   PRECONDITION(expr.type().id()==ID_pointer);
-  const pointer_typet &replacement_pointer_type=
-    pointer_type_selector.convert_pointer_type(pointer_type, ns);
+  const pointer_typet &replacement_pointer_type =
+    pointer_type_selector.convert_pointer_type(
+      pointer_type, generic_parameter_specialization_map, ns);
 
   // If we are changing the pointer, we generate code for creating a pointer
   // to the substituted type instead
-  if(replacement_pointer_type!=pointer_type)
+  // TODO if we are comparing array types we need to compare their element
+  // types. this is for now done by implementing equality operator especially
+  // for reference types, technical debt TG-2707
+  if(
+    (replacement_pointer_type != pointer_type) ||
+    (is_reference(replacement_pointer_type) && is_reference(pointer_type) &&
+     to_reference_type(replacement_pointer_type)
+       .
+       operator!=(to_reference_type(pointer_type))))
```

Comment:
Don't use `.operator!=`, just use `!=`?
Also I'm not sure how this works. Perhaps for readability it could be broken into two local booleans. None the less, I can't see how `r == p` but casting it using `to_reference_type` and then checking again can produce a different result?

Reply:
I replaced the operator with a function called `equals`